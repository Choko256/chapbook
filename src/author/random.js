/* Author unctions for random behavior. */

import seedrandom from 'seedrandom';
import shuffle from 'lodash.shuffle';

export default class {
	constructor(seed) {
		this.seed = seed;
		this.rng = seedrandom(seed);
	}

	get seed() {
		return this._seed;
	}

	/*
	Seeds the random number generator so that subsequent calls are predictable.
	Useful for testing.
	*/

	set seed(value) {
		this._seed = value;
		this.rng = seedrandom(value);
	}

	/*
	Returns either true or false, randomly.
	*/

	coinFlip() {
		return this.rng() > 0.5;
	}

	/*
	Returns a set of simulated die rolls. See
	https://en.wikipedia.org/wiki/Dice_notation for an explanation of the format
	this accepts.
	*/

	roll(spec) {
		const bits = spec.split(/d/i);

		if (bits.length !== 2) {
			throw new Error(
				'A roll must in the format [number of rolls]d[number of sides], i.e. 1d4.'
			);
		}

		const rolls = parseInt(bits[0]);

		if (isNaN(rolls)) {
			throw new Error(
				`The number of rolls specified (${bits[0]}) isn't a number.`
			);
		}

		const sides = parseInt(bits[1]);

		if (isNaN(sides)) {
			throw new Error(
				`The number of die sides specified (${bits[1]}) isn't a number.`
			);
		}

		let total = 0;

		for (let i = 0; i < rolls; i++) {
			total += 1 + Math.round(this.rng() * (sides - 1));
		}

		return total;
	}

	/*
	If only one argument is passed and it is an array:
		Returns a randomly chosen item in the array.
	Otherwise:
		Returns a randomly chosen argument.
	*/

	choice(...choices) {
		let toChooseFrom = choices;

		if (choices.length === 1 && Array.isArray(choices[0])) {
			toChooseFrom = choices[0];
		}

		return toChooseFrom[Math.round(this.rng() * (toChooseFrom.length - 1))];
	}

	/*
	If only one argument is passed and it is an array:
		Returns a copy of the array, but with items are in random order.
	Otherwise:
		Returns the arguments passed in random order.
	*/

	shuffle(...items) {
		let toShuffle = items;

		if (items.length === 1 && Array.isArray(items[0])) {
			toShuffle = items[0];
		}

		return shuffle(toShuffle.slice());
	}
}
