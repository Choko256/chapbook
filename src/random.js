/*
Functions for random behavior.
*/

const seedrandom = require('seedrandom');

module.exports = class {
	constructor(seed) {
		this.rng = seedrandom(seed);
	}

	/*
	Seeds the random number generator so that subsequent calls are predictable.
	Useful for testing.
	*/

	seed(value) {
		this.rng = seedrandom(value);
	}

	/*
	Returns either true or false, randomly.
	*/

	coinFlip() {
		return this.rng() > 0.5;
	}

	/*
	Returns a set of simulated die rolls. See
	https://en.wikipedia.org/wiki/Dice_notation for an explanation of the format
	this accepts.
	*/

	roll(spec) {
		const bits = spec.split(/d/i);

		if (bits.length !== 2) {
			throw new Error(
				'A roll must in the format [number of rolls]d[number of sides], i.e. 1d4.'
			);
		}

		const rolls = parseInt(bits[0]);

		if (isNaN(rolls)) {
			throw new Error(`The number of rolls specified (${bits[0]}) isn't a number.`);
		}

		const sides = parseInt(bits[1]);

		if (isNaN(sides)) {
			throw new Error(`The number of die sides specified (${bits[1]}) isn't a number.`);
		}

		let total = 0;

		for (let i = 0; i < rolls; i++) {
			total += 1 + Math.round(this.rng() * (sides - 1));
		}

		return total;
	}

	/*
	Returns a randomly chosen argument.
	*/

	choice(...choices) {
		return choices[Math.round(this.rng() * (choices.length - 1))];
	}
};