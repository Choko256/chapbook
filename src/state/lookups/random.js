/*
Computed variables that return rendom values each time they are read. These are
guaranteed to stay consistent over multiple sessions. They are:

-	`random.coinFlip`: either true or false
-	`random.fraction`: value 0-1
-	`random.d4`: a value 1-4
-	`random.d5`: a value 1-5
-	`random.d6`: a value 1-6
-	`random.d8`: a value 1-8
-	`random.d10`: a value 1-10
-	`random.d12`: a value 1-2
-	`random.d20`: a value 1-20
-	`random.d25`: a value 1-25
-	`random.d50`: a value 1-50
-	`random.d100`: a value 1-100
-	`random.d1000`: a value 1-1000

This also adds two regular variables: `config.random.seed`, which is the seed
the random number generator (RNG) uses, and `config.random.privateState`, which
ensures that the RNG behaves deterministically across sessions.
*/

import seedrandom from 'seedrandom';
import event from '../../event';
import {set} from '..';
import {story} from '../../story';

export const defaults = {
	'config.random.seed': () => story.name,
	'config.random.privateState': null
};

/*
Default the seed to the current time, but if it's ever changed, reset the random
number generator. This causes these random functions to behave in a
deterministic fashion.
*/

let rng = seedrandom(new Date(), {state: true});

event.on('state-change', ({name, value}) => {
	if (name === 'config.random.seed') {
		rng = seedrandom(value, {state: true});
	}
});

function saveRngState() {
	set('config.random.privateState', rng.state());
}

export default function(setComputed) {
	setComputed('random.coinFlip', () => {
		const value = rng();

		saveRngState();
		return value > 0.5;
	});

	setComputed('random.fraction', () => {
		const value = rng();

		saveRngState();
		return value;
	});

	[4, 5, 6, 8, 10, 12, 20, 25, 50, 100, 1000].forEach(i => {
		setComputed(`random.d${i}`, () => {
			const value = 1 + Math.round(rng() * (i - 1));

			saveRngState();
			return value;
		});
	});
}
