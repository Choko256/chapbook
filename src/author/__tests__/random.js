import event from '../../event';
import random from '../random';

describe('random', () => {
	test('returns boolean values from coinFlip()', () => {
		expect(typeof random.coinFlip()).toBe('boolean');
	});

	test('rolls dice with roll()', () => {
		let result = random.roll('1d4');

		expect(result >= 1 && result <= 4).toBe(true);

		result = random.roll('5d20');

		expect(result >= 5 && result <= 100).toBe(true);
	});

	test('rejects incorrect specs with roll()', () => {
		expect(() => random.roll('d4')).toThrow();
		expect(() => random.roll('5')).toThrow();
		expect(() => random.roll(5)).toThrow();
		expect(() => random.roll('nd4')).toThrow();
		expect(() => random.roll('4dx')).toThrow();
	});

	test('selects args with choice()', () => {
		let result = random.choice('a', 'b', 'c');

		expect(result === 'a' || result === 'b' || result === 'c').toBe(true);
		expect(random.choice('a')).toBe('a');
	});

	test('selects from an array with choice()', () => {
		let result = random.choice(['a', 'b', 'c']);

		expect(result === 'a' || result === 'b' || result === 'c').toBe(true);
		expect(random.choice(['a'])).toBe('a');
	});

	test('shuffles args with shuffle()', () => {
		let result = random.shuffle('a', 'b', 'c');

		expect(result.length).toBe(3);
		expect(result.filter(i => i === 'a').length).toBe(1);
		expect(result.filter(i => i === 'b').length).toBe(1);
		expect(result.filter(i => i === 'c').length).toBe(1);
	});

	test('shuffles a copy of an array with shuffle()', () => {
		let src = ['a', 'b', 'c'];
		let result = random.shuffle(src);

		expect(result.length).toBe(3);
		expect(result.filter(i => i === 'a').length).toBe(1);
		expect(result.filter(i => i === 'b').length).toBe(1);
		expect(result.filter(i => i === 'c').length).toBe(1);
		expect(src[0]).toBe('a');
		expect(src[1]).toBe('b');
		expect(src[2]).toBe('c');
	});

	test('acts predictably when seeded', () => {
		event.emit('state-change', {name: 'config.random.seed', value: 1234});

		expect(random.roll('1d10')).toMatchSnapshot();
		expect(random.roll('1d10')).toMatchSnapshot();
		expect(random.roll('1d10')).toMatchSnapshot();
	});
});
