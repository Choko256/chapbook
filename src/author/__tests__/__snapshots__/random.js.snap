// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`random acts predictably when seeded 1`] = `4`;

exports[`random acts predictably when seeded 2`] = `3`;

exports[`random acts predictably when seeded 3`] = `6`;
